Error object: add type link

ethanresnick opened the pull request:
A minimal PR based on my latest comment in #683. See that first for some caveats on this. Otherwise, I think it's good to merge.

@jmeas, Github won't let me request a review from you for some reason, but you're welcome to weigh in on this too.

If anyone else has thoughts, feel free to comment here or in #683 too.

ethanresnick committed:
Error object: add type link

See #683

jamesplease commented:
Thanks for putting this together, @ethanresnick! I think adding something like this to the errors link object makes a lot of sense. I'm not as optimistic that this could obviate the need for the `code` property, but there's no question that this will nicely complement the `code`.

One concern I have is that the difference between the `about` and `type` links may come across as too subtle right now. Their similarity may make it hard for implementers to use them confidently. 

The differences, according to this PR, are:

- `about` references more information about the _specific_ occurrence of the error,`type` includes more information about the _type_ of error
- `type` must be human-readable

Limiting the conversation to just human-readable links, I've seen two types of links that I might consider using in my errors. One is just:

`thing.com/errors/the-problem-you're-facing`

These sorts of errors are usually human-readable, yet generic, in that the same message is displayed for all occurrences of the issue. So based on my reading of the spec, these fall under `type`. Another type of message I've seen is:

`thing.com/errors/the-problem-you're-facing?component=Table&method=doSomething`

This is a templated human-readable description of the problem that uses the query params to provide a detail specific to the occurrence of the problem. The fact that this is human-readable suggests that this might be a `type`, but considering that it's been tailored to the specific occurrence of the problem, it seems that this is better suited as an `about` link.

I may be misunderstanding the difference, or perhaps we would want both of these to fall under `type` links?

Either way, I wonder if we can emphasize the difference between these two links more in the spec. Perhaps through examples, or comparing/contrasting the two?

dgeb commented on _format/1.1/index.md:
```diff
@@ -1803,6 +1803,9 @@ An error object **MAY** have the following members:
 * `links`: a [links object][links] containing the following members:
   * `about`: a [link][links] that leads to further details about this
     particular occurrence of the problem.
+  * `type`: a [link][links] whose URI that identifies the type of error that 
```
> whose URI that identifies

Use of `whose` + `that` is a little awkward. Could we just strike "whose URI" entirely?

ethanresnick commented:
@jmeas @dgeb Hey guys, sorry I haven't responded here (after, I know, being very anxious to make this change happen). I think James raised some really good points that I want to think about more, but I'll try to get back to this ASAP

ethanresnick committed:
Error object: clarify type vs about links; allow type to take an array of links

ethanresnick commented:
Hey all! Finally following up on this, and hoping we can get it merged.

To James' questions:

> The differences, according to this PR, are:
> - `about` references more information about the specific occurrence of the error, `type` includes more information about the type of error
> - type must be human-readable

The first bullet point is meant to be the only difference. In other words, `type` doesn't *need* to be dereferencable to a human readable description (the PR makes that a SHOULD). Conversely, `about` should point to a human-readable description just like `type`. I've updated the PR to add that "SHOULD" to about, so that it's now clearer that the only difference is: type = class of error, about = instance of error. 

Given that, you'd have:

```json
{ 
  "type": "thing.com/errors/the-problem-you're-facing",
  "about": "thing.com/errors/the-problem-you're-facing?component=Table&method=doSomething"
}
```

I've also updated this PR to have `type` be our first link that can hold an array of links. We're definitely going to need links that can take an array (e.g., for `profile`) and we already called out in 1.0 that we'd add links like this in the future, so I think this is a fully compatible addition that implementations have been adequately warned about:

> Note: Additional members may be specified for links objects and link objects in the future. It is also possible that the allowed values of additional members will be expanded (e.g. a collection link may support an array of values, whereas a self link does not).

ethanresnick commented:
@dgeb I think we were all on board about this a while back, but I was hesitant to merge it for various reasons. I'd like to merge it now, though, so that profiles can define their own recognizable errors (e.g. "filter using invalid operator") and identify them by unique `type` urls (which would be sub paths of the profile's main url; e.g. `` `${profile_url}/errors/invalid-operator` ``), rather than with `code` values that might conflict. Are you still game to merge this?

ethanresnick committed:
Merge branch 'gh-pages' into error-uris

dgeb commented on _format/1.1/index.md:
```diff
@@ -2299,8 +2300,12 @@ An error object **MAY** have the following members:
+  * `about`: a [link][link] that leads to further details about this
+    particular occurrence of the problem. When derefenced, this URI **SHOULD**
+    return a human-readable description of the error.
+  * `type`: a [link][link] or array of links that identifies the type of error 
```
Why allow an array of links for `type` but not for other links?

ethanresnick replied:
Allowing an array for `about` and other existing links would be a breaking change (probably + unfortunately), but `type` should be an array because an error can obviously be of multiple types, and there's precedent for array-shaped links with `profile`. Ideally, I'd like to have all new links going forward hold an array for consistency

dgeb replied:
I am not sure it's a benefit to allow all links to hold an array - consider the `self` link for instance.

And there is a singular aspect to defining each error in an error object. `code` is singular as are the other links. Furthermore, `type` throughout the spec is used to define a singular type of resource, so it seems a nice parallel to define a singular error link. Having a single link to turn to for info can be a nice simplification.

Perhaps I'm missing something. I'm definitely open to hearing some use cases that make you think an array here would be desirable.

dgeb replied:
Having discussed this a bit more with @ethanresnick, I am now ok with allowing a singular link or array of links here. The array could be used for further refinements of a `type` - from general to specific. This could allow error objects in an API to evolve while still supporting old clients looking for a particular link.

dgeb commented on _format/1.1/index.md:
```diff
@@ -536,9 +536,10 @@ Within this object, a link **MUST** be represented as either:
+Except for the `profile` key in the top-level links object and the `type` 
+key in an [error object]'s links object, each key present in a links object 
+**MUST** have a single link as its value. The aforementioned `profile` and
+`type` keys, if present, **MUST** hold an array of links.
```
This phrasing has gotten a bit awkward, and is not entirely consistent with error `type` (which can contain a singular link or an array). I'd prefer defining `profile`'s requirements separately and then using general wording such as "Except as explicitly noted elsewhere in this specification, each key present in a links object **MUST** have a single link as its value."

ethanresnick replied:
Thanks for catching this! I totally didn't realize that I was saying `type` could be a single link in one place, but that it had to be an array in another place.

If we allow `type` to be a singular link or an array, we end with two formats that clients have to support (array and object), which is more total complexity than just having one format, for no real benefit. So I'd prefer requiring `type` to be an array since, of course, a server can always send an array of one item.

dgeb replied:
Yeah, that's fine by me - and is consistent with `profile`. I suppose when we allow an array, we should just require it for simplicity.

Just to reiterate: I really don't want all new links we introduce to be arrays by default - sometimes there's value in a single canonical link (like `self`)

ethanresnick replied:
> Yeah, that's fine by me - and is consistent with `profile`.

Awesome.

> I suppose when we allow an array, we should just require it for simplicity... Just to reiterate: I really don't want all new links we introduce to be arrays by default - sometimes there's value in a single canonical link (like self)

Yup, agreed. That is, if array is allowed, it's required, but there could be links where an array isn't allowed (like `self`). 

I think the number of link types that are truly singular is pretty small, tbh. So, if we were starting from scratch, I'd say to have only one format (arrays), and then enforce that certain link types (`self`) can only hold one link through a normative MUST. But that ship has sailed, and there's (probably) not an elegant, backwards-compatible way to get there.

ethanresnick committed:
fix(errors): clarify that type link must be an array

wimleers commented:
I think it'd be great to have this crystal clear example from 
https://github.com/json-api/json-api/pull/1199#issuecomment-375452664 in the spec itself:
```
{ 
  "type": "thing.com/errors/the-problem-you're-facing",
  "about": "thing.com/errors/the-problem-you're-facing?component=Table&method=doSomething"
}
```

ethanresnick commented:
@wimleers Sure; please open a PR! Although note that type is now an array, so it'd be:

```
{ 
  "type": ["thing.com/errors/the-problem-you're-facing"],
  "about": "thing.com/errors/the-problem-you're-facing?component=Table&method=doSomething"
}
```
